### PR TITLE
copy additional files to generated app-folder (deep-copy)

### LINKED ISSUE #83: how to copy external resources to native/{project_name}/app directory
Hello,
This plugin is great, but one problem I encountered is I couldn't copy the external resources directory under my project root folder to "target/jfx/native/{project_name}/app" when I run "mvn clean jfx:native", I have tried copy-resources plugin, but it's not work.
Below is some pieces of my pom.xml configuration.

	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<jdk.version>1.8</jdk.version>
		<junit.version>4.12 </junit.version>
	</properties>

	<dependencies>
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>${junit.version}</version>
			<scope>test</scope>
		</dependency>
	</dependencies>

	<build>
		<finalName>${artifactId}</finalName>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.1</version>
				<configuration>
					<source>${jdk.version}</source>
					<target>${jdk.version}</target>
					<showWarnings>true</showWarnings>
				</configuration>
			</plugin>
			<plugin>
				<groupId>com.zenjava</groupId>
				<artifactId>javafx-maven-plugin</artifactId>
				<version>8.1.2</version>
				<configuration>
					<mainClass>com.liudonghua.apps.signapk_fx.MainApp</mainClass>
					<verbose>true</verbose>
				</configuration>
			</plugin>

			<plugin>
				<artifactId>maven-resources-plugin</artifactId>
				<version>2.7</version>
				<executions>
					<execution>
						<id>copy-resources-1</id>
						<phase>validate</phase>
						<goals>
							<goal>copy-resources</goal>
						</goals>
						<configuration>
							<outputDirectory>${basedir}/target/resources</outputDirectory>
							<resources>
								<resource>
									<directory>resources</directory>
									<filtering>false</filtering>
								</resource>
							</resources>
						</configuration>
					</execution>
					<execution>
						<id>copy-resources-2</id>
						<phase>validate</phase>
						<goals>
							<goal>copy-resources</goal>
						</goals>
						<configuration>
							<outputDirectory>${basedir}/target/jfx/native/signapk_fx/app/resources</outputDirectory>
							<resources>
								<resource>
									<directory>resources</directory>
									<filtering>false</filtering>
								</resource>
							</resources>
						</configuration>
					</execution>
				</executions>
			</plugin>

		</plugins>
	</build>

Hope someone can help me with this problem.
Thanks！

### PR DESCRIPTION
**Pull request by FibreFoX**
Fixes #83

### CODE REVIEW

**Comment by FibreFoX** [comment]
Hi @maxd, what do you think about this?

File: src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java [hunk 1]
```diff
@@ -244,6 +257,43 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+                        public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes attrs) throws IOException {
+                            // do create subfolder (if needed)
+                            Files.createDirectories(jfxAppOutputDir.toPath().resolve(additionalAppResources.toPath().relativize(path)));
+                            return FileVisitResult.CONTINUE;
+                        }
+
+                        @Override
+                        public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
+                            // do copy
+                            java.nio.file.Files.copy(path, jfxAppOutputDir.toPath().resolve(additionalAppResources.toPath().relativize(path)), StandardCopyOption.REPLACE_EXISTING);
```
Comment (maxd) [inline src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java]: Redundant `java.nio.file` package?

File: src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java [hunk 2]
```diff
@@ -188,6 +193,14 @@
      */
     protected boolean jfxCallFromCLI;
 
+    /**
+     * When you need to add additional files to generated app-folder (e.g. README, license, third-party-tools, ...),
+     * you can specify the source-folder here. All files will be copied recursively.
+     * 
+     * @parameter
+     */
+    protected File additionalAppResources;
```
Comment (maxd) [inline src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java]: Should this parameter has default value?

File: src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java (hunk 1 above)
Comment (maxd) [inline src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java]: I think need to specify COPY_ATTRIBUTES copy option.

File: src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java [hunk 3]
```diff
@@ -244,6 +257,43 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+                        public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
+                            // do copy
+                            java.nio.file.Files.copy(path, jfxAppOutputDir.toPath().resolve(additionalAppResources.toPath().relativize(path)), StandardCopyOption.REPLACE_EXISTING);
+                            return FileVisitResult.CONTINUE;
+                        }
+
+                        @Override
+                        public FileVisitResult visitFileFailed(Path path, IOException ioe) throws IOException {
+                            // don't fail, just inform user
+                            getLog().info(String.format("Couldn't copy additional app resource %s", path.toString()));
```
Comment (maxd) [inline src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java]: 1. How about warning (not info)?
1. `path` is source or destination path?

File: src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java [hunk 4]
```diff
@@ -244,6 +257,43 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+                        }
+
+                        @Override
+                        public FileVisitResult postVisitDirectory(Path path, IOException ioe) throws IOException {
+                            // nothing to do here
+                            return FileVisitResult.CONTINUE;
+                        }
+                    });
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (maxd) [inline src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java]: I think need to complete execution with not zero exit code. What do you think?

**Comment by maxd** [comment]
Seems very useful :+1: . I will try to migrate coping LICENSE file in my project to functionality from this PR this week for test it.

**Comment by FibreFoX** [comment]
Hi @maxd, thanks again for reviewing my changes.
I will adjust this PR by removing redundant ```java.nio```-package-usage and change from INFO to WARNING, and will rename ```path``` to proper names.
About non-zero return: As we are able to run inside maven-lifecycle now, I fear to break built to CI-projects. Having notification as WARNING instead of INFO should be enough for developers to find the problem. The idea behind this: "don't fail the build because of file-permission-problems on the build-system".

The new configuration-parameter does not need any default, we already have src/main/deploy.... and other conventions which are easily overseen (mostly because of missing proper documentation).

**Comment by maxd** [comment]
Seems ok. My LICENSE file was copied to `jfx/app` directory.

**Comment by FibreFoX** [comment]
Wonderful, this will get into our new website, because most of the questions are "how to add license" and "how to customize native bundle". I will merge this now.

**Pull request merged**